tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java

```diff
@@ -1,23 +1,124 @@
+  private final Paginator paginator;
 
   @GetMapping(path = "/events")
   @ResponseStatus(HttpStatus.OK)
-  public List<EventTO> findAllEvents() {
-    return eventService.findAll();
+  public List<EventTO> findEvents(
+    @RequestParam(value = "eventType", required = false)
+    @EnumSubset(anyOf = {"SHIPMENT", "TRANSPORT", "EQUIPMENT"})
+    String eventType,
+
+    @RequestParam(value = "shipmentEventTypeCode", required = false)
+    ShipmentEventTypeCode shipmentEventTypeCode,
+
+    @RequestParam(value = "carrierBookingReference", required = false) @Size(max = 35)
+    String carrierBookingReference,
+
+    @RequestParam(value = "transportDocumentReference", required = false) @Size(max = 20)
+    String transportDocumentReference,
+
+    @RequestParam(value = "transportEventTypeCode", required = false)
+    TransportEventTypeCode transportEventTypeCode,
+
+    @RequestParam(value = "transportCallID", required = false) @Size(max = 100)
+    String transportCallID,
+
+    @RequestParam(value = "vesselIMONumber", required = false) @Size(max = 7)
+    String vesselIMONumber,
+
+    @RequestParam(value = "carrierVoyageNumber", required = false) @Size(max = 50)
+    String carrierVoyageNumber,
+
+    @RequestParam(value = "carrierServiceCode", required = false) @Size(max = 5)
+    String carrierServiceCode,
+
+    @RequestParam(value = "equipmentEventTypeCode", required = false)
+    EquipmentEventTypeCode equipmentEventTypeCode,
+
+    @RequestParam(value = "equipmentReference", required = false) @Size(max = 15)
+    String equipmentReference,
+
+    @RequestParam(value = "limit", defaultValue = "100", required = false) @Min(1)
+    int limit,
+
+    @RequestParam(value = "API-Version", required = false)
+    String apiVersion,
+
+    HttpServletRequest request, HttpServletResponse response
+  ) {
+    Cursor cursor = paginator.parseRequest(
+      request,
+      new CursorDefaults(limit, Sort.Direction.ASC, EventCache_.EVENT_CREATED_DATE_TIME)
+    );
+
+    PagedResult<EventTO> result = eventService.findAll(cursor, EventCacheFilters.builder()
+        .eventType(toEnumList(eventType, EventType.class))
+        .shipmentEventTypeCode(shipmentEventTypeCode)
+        .carrierBookingReference(carrierBookingReference)
+        .transportDocumentReference(transportDocumentReference)
+        .transportEventTypeCode(transportEventTypeCode)
+        .transportCallID(transportCallID)
+        .vesselIMONumber(vesselIMONumber)
+        .carrierVoyageNumber(carrierVoyageNumber)
+        .carrierServiceCode(carrierServiceCode)
+        .equipmentEventTypeCode(equipmentEventTypeCode)
+        .equipmentReference(equipmentReference)
+      .build());
+
+    paginator.setPageHeaders(request, response, cursor, result);
+    return result.content();
+  }
+
+  /**
+   * Splits a comma-separated string and converts each subpart into
+   * an enum.
+   */
+  private <T extends Enum<T>> List<T> toEnumList(String values, Class<T> enumClass) {
```

Comment:
fine for now, but possible candidate to move to shared kernel?

Reply:
Yeah I suspect I will move in when implementing event-subscriptions as I might need it there as well